### PR TITLE
Implement duels

### PR DESCRIPTION
**Pull request by ubeF**
Implement duel scene:
- Coroutine:
  - Params: two players - attacker and target
  - Return: signal, which resolves to either Win, Loss or Draw (use Enum)
- Use question difficulty dialog to get difficulty
- ask up to 3 questions
- every question has to be answered 2 times, once by each player
- display which player answers when
- only ask next question, if question before that was a draw
- if all 3 questions are draw return Draw
- announce result
- emit signal with result

### CODE REVIEW

**Comment by ubeF** [comment]
Not sure if the files in duel/ are of any help, i just moved them here

**Comment by KevinBasener** [comment]
Implemented requested features and added players name, so it is possible to differentiate whos turn it is.

File: main/main.gd [hunk 1]
```diff
@@ -71,6 +71,16 @@ func create_players(num):
...
 
+		match i: 
+			0:
+				player.player_name = "Euro"
+			1:
+				player.player_name = "Yen"
+			2:
+				player.player_name = "Pound"
+			3:
+				player.player_name = "Dollar"
```
Comment (hfxbse) [inline main/main.gd]: Creating an array for the names like it is done with the sprites seems more maintainable IMO.

File: player/player.gd [hunk 2]
```diff
@@ -121,12 +122,29 @@ func register_duels():
...
+	var result = await duel_panel.start_duel(player, self)
+	match result:
+		duel_panel.DuelEndings.ATTACKER_WINS:
+			var pot = self.riesen / 2
+			self.riesen -= pot
+			player.riesen += pot
+		duel_panel.DuelEndings.ATTACKER_LOSES:
+			var pot = player.riesen / 2
+			self.riesen += pot
+			player.riesen -= pot
```
Comment (hfxbse) [inline player/player.gd]: Not a fan of the match statement here. Why not do it like this to not repeat yourself?

```gdscript
# TODO: Add an early exit if it is possible to draw
var pot = self.riesen / 2

if result == dual_panel.DualEndings.ATTACKER_LOSES:
    pot *= -1

self.riesen += pot
player.riesen -= pot
```

As you have the same match statement further down in your code, create a separate function for it. Do not duplicate code. I image the function something like this:

```gdscript
func apply_dual_outcome(other_player: Player, bet_amount: int: result: DualEndings):
    ...
```

File: player/player.gd [hunk 3]
```diff
@@ -121,12 +122,29 @@ func register_duels():
 	
 	
 func duel(player: Player):
-	print("Duel triggered")
+	var duel_panel = preload("res://duel/duell_movement/same_field_duel.tscn").instantiate()
+	add_child(duel_panel)
+	var result = await duel_panel.start_duel(player, self)
+	match result:
+		duel_panel.DuelEndings.ATTACKER_WINS:
```
Comment (hfxbse) [inline player/player.gd]: Might be worth it to access the enum through the class name instead of the instance.

File: question_panel/answer_button.gd
Comment (hfxbse) [inline question_panel/answer_button.gd:1]: Only made the formating worse, please revert it to the version in main.

File: question_panel/question_panel.gd [hunk 4]
```diff
@@ -2,19 +2,38 @@ extends CanvasLayer
 
 var question: Question
 var selected_answer
-
+
```
Comment (hfxbse) [inline question_panel/question_panel.gd]: Unnecessary tab.

File: question_panel/question_panel.gd [hunk 5]
```diff
@@ -2,19 +2,38 @@ extends CanvasLayer
...
 var selected_answer
-
+	
 @onready var question_label = $MarginContainer/QuestionContainer/QuestionLabel
 @onready var answer_container = $MarginContainer/QuestionContainer/AnswerContainer
 @onready var question_container = $MarginContainer/QuestionContainer
 
+@export var player_name: String = "":
+	set(text):
+		if question_label:
```
Comment (hfxbse) [inline question_panel/question_panel.gd]: You should also check if the value is actually different than the previous one.

File: question_panel/question_panel.gd [hunk 6]
```diff
@@ -2,19 +2,38 @@ extends CanvasLayer
...
 
 func display_question(difficulty):
 	var question_handler = preload("res://data_providers/question_provider/question_provider.tres")
 	
 	question = question_handler.get_random_question(difficulty)
-	question_label.set_text(question.question)
+	question_label.text = question_label.text + question.question
+
+	for answer_possibility in question.answer_possibilities:
+		var answer_button = create_answer_button(answer_possibility, _on_answer_selected)
```
Comment (hfxbse) [inline question_panel/question_panel.gd]: Not really related to this PR, but IMO you should not be forced to add a receiver to the button during the creation. Just add it directly on it in a second line.

```gdscript
var answer_button = create_answer_button(answer_possibility)
answer_button.connect(_on_answer_selected)
```

File: question_panel/question_panel.gd [hunk 7]
```diff
@@ -2,19 +2,38 @@ extends CanvasLayer
...
+		
+		answer_container.add_child(answer_button)
+
+	return answer_selected
+	
+func display_specific_question(question: Question):
+	self.question = question
+	
+	var question_handler = preload("res://data_providers/question_provider/question_provider.tres")
```
Comment (hfxbse) [inline question_panel/question_panel.gd]: The variable `question_handler`, which should be called `question_provider` in the first place, is never used. This smells like copy 'n pasta, which it is actually.

Convert it to this instead:
```gdscript
func display_question(difficulty: Question.Difficulty):
	var question_handler = preload("res://data_providers/question_provider/question_provider.tres")
	display_specific_question(question_handler.get_random_question(difficulty))

func display_specific_question(question: Question):
    ...
```

**Review by hfxbse — CHANGES_REQUESTED** [review]
Gameplay-wise: Why not reveal the answer to the question after both players answered instead? Would be cool to show both logged-in answers on the same panel but I believe we have no time left to do so.

The question title seems to be overflowing, is auto-wrapping enabled? I also don't it is necessary to prefix the question with "Question:"

**Comment by KevinBasener** [comment]
All issues adressed

File: player/player.gd (hunk 2 above)
Comment (ubeF) [reply player/player.gd]: What about a draw. In that case nothing is supposed to happen

**Pull request closed without merging**